### DISCUSSION TITLE
BoundaryConditions::constant_exterior's purpose?

### DISCUSSION THREAD

**Discussion by icebeing** [halide/Halide#7993] [Q&A] [answered]
Hi all -

   So, I've been looking over the above helper Func.  Would this be a function object that effectively constrains a buffer's contents to only exist within a certain sub-rectangle?

   Let's say I have the following:  
     - an input buffer whose rows are aligned to some multiple of 128
     - the input buffer's stride > width
     - in the bytes beyond the width within each row, is garbage data (noise)

   To ignore the noise part of the input, I can define this Func object:

    Input<Buffer<uint8_t>, 2> strided_input;
    Func my_true_image = BoundaryConditions::constant_exterior(strided_input, 0, {0, strided_input.width()}, {0, strided_input.height()});

     How would the realization of this stage behave, especially when vectorization is involved?  When we encounter the pixels beyond the Region definition, will the code automagically pad the overlapped bytes with zeros?

    I guess a side-question:  can strided_input have it such that the width & stride are different values?  

I think the ./test/correctness/boundary_conditions.cpp Unit-test hints at the answer, but I just wanted to know from someone that's used this function before.  

Thanks, Charles.

**Comment by abadams** [comment, marked as answer]
I wouldn't say it constrains it to only exist within a sub-rectangle. I'd say it is a Func that equals the first arg (a Func) within some subrectangle, and equals the second arg (a constant) outside of that subrectangle.

Typically when you vectorize this Halide will split the loop over vectors into three parts: a prologue for vectors that are before or overlapping the left edge, a steady-state for vectors entirely within the Func arg, and an epilogue for vectors that are after or overlapping the right edge.

Compiling to stmt_html and reading the result might be illuminating.
